libs/ledger-live-common/src/hw/actions/implementations.ts

```diff
@@ -0,0 +1,286 @@
+import {
+  EMPTY,
+  Observable,
+  Observer,
+  ReplaySubject,
+  Subscription,
+  TimeoutError,
+  concat,
+  of,
+  timer,
+} from "rxjs";
+import {
+  DisconnectedDevice,
+  DisconnectedDeviceDuringOperation,
+} from "@ledgerhq/errors";
+import {
+  catchError,
+  debounce,
+  filter,
+  switchMap,
+  tap,
+  timeout,
+} from "rxjs/operators";
+import { Device } from "./types";
+import { getDeviceModel } from "@ledgerhq/devices";
+import { ConnectManagerTimeout } from "../../errors";
+
+export enum ImplementationType {
+  event = "enum",
+  polling = "polling",
+}
+
+type ImplementationEvent =
+  | {
+      type: "deviceChange";
+      device?: Device | undefined | null;
+    }
+  | {
+      type: "unresponsiveDevice";
+    }
+  | {
+      type: "error";
+      error: Error;
+    };
+
+type PollingImplementationConfig = {
+  pollingFrequency: number;
+  initialWaitTime: number;
+  reconnectWaitTime: number;
+  connectionTimeout: number;
+};
+
+type PollingImplementationParams<Request, EmittedEvents> = {
+  deviceSubject: ReplaySubject<Device | null | undefined>;
+  task: (params: {
+    deviceId: string;
+    request: Request;
+  }) => Observable<EmittedEvents>;
+  request: Request;
+  config?: PollingImplementationConfig;
+};
+
+const defaultConfig: PollingImplementationConfig = {
+  pollingFrequency: 2000,
+  initialWaitTime: 5000,
+  reconnectWaitTime: 5000,
+  connectionTimeout: 20000,
+};
+type Implementation = <EmittedEvent, GenericRequestType>(
+  params: PollingImplementationParams<GenericRequestType, EmittedEvent>
+) => Observable<EmittedEvent | ImplementationEvent>;
+
+export type EmittedEvent<T> = {
+  type: string;
+} & T;
+
+/**
+ * Returns a polling implementation function that repeatedly performs a given task
+ * with a given request, with a specified polling frequency and wait times until a
+ * KO or OK end state is reached. This emulates the event based connection paradigm
+ * from the USB stack in a BLE setting.
+ * @template EmittedEvent - Type of the events emitted by the task's observable.
+ * @template GenericRequestType - Type of the request to be passed to the task.
+ **/
+const pollingImplementation: Implementation = <
+  SpecificType,
+  GenericRequestType
+>(
+  params: PollingImplementationParams<GenericRequestType, SpecificType>
+): Observable<SpecificType | ImplementationEvent> =>
+  new Observable(
+    (
+      subscriber: Observer<EmittedEvent<SpecificType> | ImplementationEvent>
+    ) => {
+      const { deviceSubject, task, request, config = defaultConfig } = params;
+      const {
+        pollingFrequency,
+        initialWaitTime,
+        reconnectWaitTime,
+        connectionTimeout,
+      } = config;
+
+      let shouldStopPolling = false;
+      let connectSub: Subscription;
+      let loopTimeout: NodeJS.Timeout | null;
+      let firstRound: boolean = true;
```

Comment:
TS says: "Type boolean trivially inferred from a boolean literal, remove type annotation" -> i guess it's trivial for TS ahah

Reply:
Damn it typescript